### PR TITLE
feat(legacy): Implement EIP-712 signing

### PR DESCRIPTION
**Pull request by alisinabh**
Hey :wave: 

In this PR I have implemented a basic EIP-712 typed data signing for T1 based on Ledger. In this approach, The client calculates sha-3 of the `domain separator` and and `message` then sends them to trezor for signing. On the device, both of the signatures will be combined and prefixed with `\x19\x01` and then get signed and sent back.

This is NOT the most secure way since an attacker can send any other message and you cannot see the actual data on the device screen, on the other hand showing a large message data is not feasible on a T1's screen and having this feature which is becoming a de-facto standard is better than not having it at all I guess.

This should not compromise anything else as the client data for signature is prefixed with `\x19\x01`.

I have also made the necessary changes in `trezor/connect` and `metamask/eth-trezor-keyring` and tested them with the emulator. If you choose to keep this, I can also make push my other changes in there. 

Please share your reviews, suggestions and concerns. I really loved working on this issue.

### CODE REVIEW

**Comment by prusnak** [comment]
Thank you @alisinabh! Much appreciated!

**Comment by matejcik** [comment]
@alisinabh can you please point me to the relevant changes to _Metamask_ code?

**Comment by alisinabh** [comment]
Hi @matejcik,

Sure. [This commit](https://github.com/alisinabh/eth-trezor-keyring/commit/78a504a2af971c6740655f9a4768d44299ff4047) was all the required changes in eth-trezor-keyring. I needed to add this changes to `v0.7.0` tag for testing because the `main` branch was not compiling the metamask extension.

I think the following part needs to be done in trezor-connect (only for T1) so we would have a unified API for both T1 and TT to sign typed data.

```js
 const { domain, types, primaryType, message } =
   sigUtil.TypedDataUtils.sanitizeData(data);

 const domainSeparatorHex = sigUtil.TypedDataUtils.hashStruct(
   'EIP712Domain',
   domain,
   types,
   isV4,
 ).toString('hex');
 const hashStructMessageHex = sigUtil.TypedDataUtils.hashStruct(
   primaryType,
   message,
   types,
   isV4,
 ).toString('hex');

```

Let me know if you need anything.

**Comment by gewure** [comment]
People are having Assets stuck because of EIP-712 requirements on various plattforms.
Please review, merge & release this ASAP.

**Comment by prusnak** [comment]
Unlocked so @alisinabh can participate in the review.

File: common/protob/messages-ethereum.proto
Comment (alisinabh) [reply common/protob/messages-ethereum.proto:135]: Or even we can move the `EthereumTypedDataSignature` message to `messages-ethereum.proto` and use this type for both of the.

Although using `EthereumMessageSignature` can save a small bit of storage on T1.

Comment (matejcik) [reply common/protob/messages-ethereum.proto:135]: (sorry about the lock, I assumed that you as the author can always comment :eyes: )

Using the same message also saves a small bit on TT. The only issue with that solution is that the firmware currently being relased is using a different response -- but given that there is no host-side implementation that can use it currently, maybe we can change it.

We'll discuss this with the rest of the team.

**Comment by gewure** [comment]
@trezor have you ever heard of the Streisand-Effect?

They way Trezor handles this situation is abyssmal and prompts us to seriouly reconsider working with any Trezor product in the future.

- at first you did not recognize the communities need for the feature
- you then actively played the down the impact of the feature
- then your product-team came up with ridiculous "this does not serve the user" arguments, when already people where desperately screaming for EIP-712
- only when very serious matters where raised in the github issues, you slowly moved (Like people losing thousands of money?)
- this was basically developed FOR FREE FOR YOU - yet you act like it was a burden?  Are you people crazy or stupid?!

Thousands of Users have NFTs stuck and lose money.

... aht what does Trezor do here? You are concered about SPAM in here - censoring messages that dont bootlick your approach :-1: :angry: 

People are very mad at you - and the internet shall know that. You are acting very untrustworthy and seem to not give anything about your users. Do we need to open a reddit to discuss this?

By censoring the Githubissues you only make us more mad. We thought long about this, but decided to publicly criticice Trezor now - in order to warn other users not to buy/use your products. 

Your support is abyssmal, and future not-customers should know, and developers on github even more so. 

in anger,
signed by a 5-headed team of Devs.

File: common/protob/messages-ethereum.proto
Comment (matejcik) [reply common/protob/messages-ethereum.proto:135]: Please move `EthereumTypedDataSignature` to this file and use it instead of `EthereumMessageSignature`

**Comment by matejcik** [comment]
hello @alisinabh, can you provide some test vectors for the call? If there are some official ones that we could use, great.

Also, can you estimate how often would a normal user need to sign EIP-712 messages? Is it a one-time thing to register somewhere, or is it more like, every transfer needs an EIP-712 confirmation?

**Comment by kargakis** [comment]
Most dapps are transitioning to use EIP-712 messages so it's quite often that you get to use it (snapshot voting, OpenSea approvals, etc.)

-------- Original Message --------

**Comment by alisinabh** [comment]
Hi @matejcik 

> can you provide some test vectors for the call? If there are some official ones that we could use, great.

About the test vector I don't completely understand what you mean. Basically I have used metamask + simple web3js script to sign with typed data. I will also add corresponding tests for this functionality. I just need to go through them a bit.

> Also, can you estimate how often would a normal user need to sign EIP-712 messages? Is it a one-time thing to register somewhere, or is it more like, every transfer needs an EIP-712 confirmation?

Basically it would depend on the user. For example all transactions in zkSync use this functionality. Voting on snapshot.org also requires this functionality. It is expanding every day as it is cheaper to verify on-chain.

File: common/protob/messages-ethereum.proto
Comment (alisinabh) [reply common/protob/messages-ethereum.proto]: Done in 69b4da3800628fda21578d0902fd8d06b3ee3e13

**Comment by matejcik** [comment]
> About the test vector I don't completely understand what you mean. Basically I have used metamask + simple web3js script to sign with typed data. I will also add corresponding tests for this functionality. I just need to go through them a bit. (**Do we even have tests for T1?**)

we have tests in `tests/device_tests/ethereum` that can execute against T1

by "test vectors", i mean combinations of seed + message data + corresponding hashes + signatures by that seed. If you can generate that by importing `all all all all all all all all all all all all` seed into Metamask and producing signatures, and then compare these signatures to those produced with Trezor, that would be best.

If you write the testcase, that's a bonus, but at this point I might take that over, just paste the relevant data here.

File: python/src/trezorlib/ethereum.py
Comment (mmilata) [inline python/src/trezorlib/ethereum.py:370]: `normalize_nfc` shouldn't be used here as we're dealing with binary data, not Unicode. Probably best to declare the arguments as `bytes` instead of `AnyStr` and do hex->bytes conversion in the the caller.

**Comment by matejcik** [comment]
I reworked the layout code and added a warning at start. I'll add device tests next.

File: legacy/firmware/layout2.c
Comment (prusnak) [inline legacy/firmware/layout2.c:1258]: Maybe just `hash`?
```suggestion
                       const uint8_t *hash, uint32_t len) {
```

Comment (matejcik) [reply legacy/firmware/layout2.c]: a082b347b5ce71547b5e115c0ffc5adacdfd0fae

**Comment by alisinabh** [comment]
Sorry for the delay. Here are some vectors as you requested. I have already tested it and it generates the same signature on the T1 emulator. (I needed to use a different seed as I was not able to import this one on the emulator)

### Message
```json
{
    "domain": { "chainId": 1, "name": "Ether Mail", "verifyingContract": "0xCcCCccccCCCCcCCCCCCcCcCccCcCCCcCcccccccC", "version": "1" },
    "message": {
        "contents": "Hello, Bob!",
        "attachedMoneyInEth": 4.2,
        "from": { "name": "Cow", "wallets": ["0xCD2a3d9F938E13CD947Ec05AbC7FE734Df8DD826", "0xDeaDbeefdEAdbeefdEadbEEFdeadbeEFdEaDbeeF"] },
        "to": [{ "name": "Bob", "wallets": ["0xbBbBBBBbbBBBbbbBbbBbbbbBBbBbbbbBbBbbBBbB", "0xB0BdaBea57B0BDABeA57b0bdABEA57b0BDabEa57", "0xB0B0b0b0b0b0B000000000000000000000000000"] }]
    },
    "primaryType": "Mail",
    "types": {
        "EIP712Domain": [
            { "name": "name", "type": "string" },
            { "name": "version", "type": "string" },
            { "name": "chainId", "type": "uint256" },
            { "name": "verifyingContract", "type": "address" }
        ],
        "Group": [
            { "name": "name", "type": "string" },
            { "name": "members", "type": "Person[]" }
        ],
        "Mail": [
            { "name": "from", "type": "Person" },
            { "name": "to", "type": "Person[]" },
            { "name": "contents", "type": "string" }
        ],
        "Person": [
            { "name": "name", "type": "string" },
            { "name": "wallets", "type": "address[]" }
        ]
    }
}
```

### Address and Signer

 - Seed: `all all all all all all all all all all all all`
 - Signing with `0x73d0385F4d8E00C5e6504C6030F47BF6212736A8`

### Signature and Hashes

 - Domain separator hash: `f2cee375fa42b42143804025fc449deafd50cc031ca257e0b194a650a912090f`
 - Message hash: `eb4221181ff3f1a83ea7313993ca9218496e424604ba9492bb4052c03d5c3df8`
 - Signature `0x1d778d9ae559161f4ea57aad9135035eb7e26e5e4cf5b571c58736ee265b649b17c38730ede957efbcf7de4f30906b133a4262b9e4bb8e4ba3927a48512e3a561c`

[Here is the script](https://gist.github.com/alisinabh/96bd50fd2f9bf827544585cee40e2f7c) that I've used if you need to change anything.

**Comment by matejcik** [comment]
> **EDIT: I just saw you already have one. Ignore this please.**

nah, i'll add it. the more vectors, the merrier :)

can you please open a PR to connect?

**Comment by matejcik** [comment]
squashed & rebased, added changelog, merged `sign_typed_data.json` and `sign_typed_data_blind.json` into a single file (in the future we hopefully gain the ability to cross-check that the partial hashes match the message).

@mmilata one final round of review pls?

**Review by mmilata — APPROVED** [review]
LGTM! Big thanks for your contribution @alisinabh!

Was thinking that maybe having couple of known-invalid test vectors would be nice. But there are only 2 i think, non-ethereum `address_n` and hash of wrong length which are probably not that interesting.

**Comment by matejcik** [comment]
waiting for gitlab CI success before merging https://gitlab.com/satoshilabs/trezor/trezor-firmware/-/pipelines/431980539

**Comment by alisinabh** [comment]
> can you please open a PR to connect?

Yes. Will do it in the weekend.

> Big thanks for your contribution @alisinabh!

My pleasure. This is why I have chosen an open source wallet. :smiley:

**Comment by prusnak** [comment]
The CI pipeline is green 🎉  https://gitlab.com/satoshilabs/trezor/trezor-firmware/-/pipelines/432051154

**Comment by matejcik** [comment]
and we're done!
thanks @alisinabh for your work!

**Comment by rsimonton** [comment]
Great work all!  Is there a target release date for this?

**Comment by matejcik** [comment]
this is slated for release on January 19.
however, you should also watch https://github.com/trezor/connect/pull/986 and a corresponding PR for Metamask, because this feature won't help you until those are in as well

**Comment by svbloot** [comment]
imagine implementing this feature for Trezor Model T in April 2021 and then just ignoring Trezor One for more than half a year. shameful support.

**Pull request merged**
